### PR TITLE
Rewrite the current App using React

### PR DESCRIPTION
**Pull request by kitten**
This just adds the basic infrastructure for React. It at the time implements just one view in the React Router called 'Map', which renders the LocalMap component.

The database reducer has to be adapted to utilise Immutable.js

### CODE REVIEW

File: src/dispatcher.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import {
+  createDispatcher
+} from 'fluorine-lib'
+
+export default createDispatcher()
```
Comment (rgbkrk) [inline src/dispatcher.js:5]: Ah, this comes from fluorine. I presume `fluorine` was taken on npm then?

**Comment by rgbkrk** [comment]
I'm getting errors on `npm run compile` for `invariant` on this branch at the moment.

**Comment by rgbkrk** [comment]
I've added in `npm run compile` before as a `prebuild`, to verify that the transpile goes through and completes.

File: gulpfile.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+/*eslint-disable */
+var gulp = require('gulp')
+var gutil = require('gulp-util')
+
+var webpack = require('webpack')
+var webpackConfig = require('./webpack.config')
+
+gulp.task('webpack', function() {
```
Comment (nelix) [inline gulpfile.js:8]: I think just running webpack might be simpler, it comes with a cli which you could add to package.json/scripts. 

``` js
"build": "webpack",
```

One less dependency?

File: package.json [hunk 3]
```diff
@@ -29,15 +31,28 @@
   },
   "homepage": "https://github.com/rgbkrk/pipboy#readme",
   "devDependencies": {
-    "babel-preset-es2015": "^6.1.18",
-    "electron-compile": "^1.0.1",
-    "electron-compilers": "^1.0.1",
+    "babel-core": "^5.8.34",
+    "babel-loader": "^5.4.0",
```
Comment (nelix) [inline package.json]: Are there plans to use any other loaders?

``` js
"compile": "babel src --out-dir app/ --source-maps",
```

File: src/app.jsx [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+  sendCommand = (type, ...args) => {
+    if (this.connection) {
+      this.connection.onNext([type, ...args])
+    }
+  }
+
+  componentWillMount() {
+    // TODO: Implement a server selection
+    discover()
```
Comment (nelix) [inline src/app.jsx:65]: I would like to see some separation between the pipboy api and the app.
But prolly not yet, just thinking.
Maybe we could pass the subjects into the views?

Comment (rgbkrk) [inline src/app.jsx:65]: > Maybe we could pass the subjects into the views?

That's a good idea. I've been wondering if we can pass the subject onto all the views, including those that get created by others. We provide the `subject` for others to subscribe to, rendering whatever they want in their view.

File: gulpfile.js (hunk 2 above)
Comment (kitten) [inline gulpfile.js:8]: @nelix Agreed

File: package.json (hunk 3 above)
Comment (kitten) [inline package.json]: Well there is the possibility of adding loaders for icons (svg), map tiles (png), stylesheets, the html, etc. So let's not remove webpack just yet ;)

File: src/app.jsx (hunk 4 above)
Comment (kitten) [inline src/app.jsx:65]: Well I'm planning to write some command actions that will be injected into the subject, thus complying to the Flux scheme

File: src/dispatcher.js (hunk 1 above)
Comment (kitten) [inline src/dispatcher.js:5]: Correct. It's currently at version 0.2.0

**Comment by kitten** [comment]
@rgbkrk We can discuss the Flux pattern on gitter if you'd like?

I think we should remove gulp for now and I should write some actions for sending commands. Sounds good, @nelix?

**Comment by rgbkrk** [comment]
I say just get this in fit shape to run, then we'll merge and we can keep on iterating. Right now, I haven't been able to adapt this to work currently.

File: src/components/LocalMap.jsx [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.map !== this.props.map) {
+      this.updateMap(nextProps.map)
+    }
+  }
+
+  updateMap = map => {
+    // Get next LocalMap
+    this.context.sendCommand('RequestLocalMapSnapshot')
```
Comment (rgbkrk) [inline src/components/LocalMap.jsx:66]: Since `updateMap` gets called in `componentDidMount`, this only requests the localmap when in view. Neat.

**Comment by kitten** [comment]
@rgbkrk @nelix Sorry for being so slow lately :( I removed webpack and gulp for now. I think this should be merged, so we can further iterate on this.

**Comment by rgbkrk** [comment]
We can't currently merge this because of the merge conflict unless we do it by hand. This also isn't passing the tests, which right now are just a sanity check on builds.

**Comment by rgbkrk** [comment]
I've brought over most of this into #14. It uses electron-compile for the moment still.

A few things got lost in translation, namely arrows.

I can't tell if it's Fallout or pipboy here but my game has been crashing hard with this local map up while at the Institute. Relay interference? :wink:

**Pull request closed without merging**